### PR TITLE
Changes assoc so it does not flatten object prototype properties into…

### LINKED ISSUE #1697: about assoc prototype flattening
As is documented, assoc flattens prototype properties into the new object.

I'm using Ramda in a MeteorJS project. I have objects that have helpers which are assigned as properties in the prototype and a schema validation that checks the structure of the object before saving it to the database. Whenever I do an operation in an object using assoc it becomes invalid because of the prototype properties flattening.

I know this is not really a Ramda issue, but I think this behavior is not really what you would expect when you're just trying to change a property in an object.

Would the performance cost be too high if we were to set the prototype properties on the new object's prototype or is there something I'm missing?

### PR DESCRIPTION
**Pull request by VitorHP**
… the new object. Fixes #1697 

Implemented a version of assoc the does not flatten prototype properties.

It runs 2-3x slower than the original one however it's still pretty fast. 

There's some measuring [here](https://jsfiddle.net/wuqq57mk/2/)

### CODE REVIEW

**Comment by asaf-romano** [comment]
Functionality-wise, I think this change is desired (see previous discussion on https://github.com/ramda/ramda/issues/1582), but only if it's applied across-the-board for all cloning operations. If we can abstract the implementation so that it's consolidated in one place, then we can also keep a fast-lane for POJOs.

**Comment by CrossEye** [comment]
I have several issues with the commit as presented, but let's discuss this in terms of _whether_ in #1697 before thinking too much about _how_.

File: src/assoc.js [hunk 1]
```diff
@@ -22,10 +21,14 @@ var _curry3 = require('./internal/_curry3');
  *      R.assoc('c', 3, {a: 1, b: 2}); //=> {a: 1, b: 2, c: 3}
  */
 module.exports = _curry3(function assoc(prop, val, obj) {
-  var result = {};
-  for (var p in obj) {
-    result[p] = obj[p];
+  var result = Object.create(Object.getPrototypeOf(obj))
+  var properties = Object.getOwnPropertyNames(obj)
```
Comment (davidchambers) [inline src/assoc.js]: `Object.create`, `Object.getPrototypeOf`, and `Object.getOwnPropertyNames` are not defined in ES3, which Ramda targets.

Comment (buzzdecafe) [inline src/assoc.js]: > ES3, which Ramda targets.

all in favor of dropping that constraint? :hand:

Comment (jdalton) [inline src/assoc.js]: `Object.getOwnPropertyNames` gets non-enumerable props. If you go down that route you'll want to use `Object.defineProperty` and `Object.getOwnPropertyDescriptor` to ensure consistency.

**Pull request closed without merging**